[deps] Update ECharts to version 5.6.0 #305

dee077 opened the pull request:
Modify import statements according to the updated ECharts module structure.

Fixes #305, #244

## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.

## Reference to Existing Issue

Closes #305 #244.

## Description of Changes

- Updated ECharts to version 5.6.
- Modified import statements to align with the updated ECharts module structure.
- Used `import "echarts-gl"` instead of the file `lib/js/eharts-gl.min.js` which was causing some issues and can be removed if needed.
- One way to do this `import * as echarts from "echarts"` will import the whole library, I have used `import * as echarts from "echarts/core"`; and added other necessary imports accordingly.
- I will add a selenium test for #269 to check the correct renders and console errors during runtime to catch bugs.

nemesifier reviewed (commented):
Encouraging! I did a quick test and seems to work! Give me a bit more time to review and test more carefully, thanks! :pray: :+1:

nemesifier reviewed (changes requested):
If you can, please add a basic test to ensure echarts is working, the test should fail with a clear error message if we execute the tests without the changes in `src/js/netjsongraph.render.js`.

nemesifier reviewed (changes requested):
During testing I noticed this:

![echarts-update](https://github.com/user-attachments/assets/cb2fec62-cc98-498b-be63-8ea5f586cd44)

Try passing the mouse over the nodes and links, something happened to the positioning of these elements and the result is not readable.

dee077 commented:
## Changes

- Increased Line height from `5` to `20` which solves the tooltip elements overlap issue.
- Added support for the selenium test.
- For now added a single test for Geographic map example and checked the following conditions:
  - Check for `leaflet container` dev exists.
  - Check for `canvas` HTML element exists.
  - Check there are no console errors except `favicon.ico` not found
  - Check the `rendered nodes and links` count from `sidebar` and the `nodes and links present` in the `netjsonmap.json` are the same
 - Will add more tests for the other examples of  map and graph once this is approved.

nemesifier commented on test/browser.test.utils.js:
```diff
@@ -0,0 +1,96 @@
+  }
+};
+
+export const getRenderedNodesAndLinks = async (driver) => {
```
does this return the count of nodes and links or the references to the nodes and links?
Seems only the count to me, I think the name could be misleading. If it returns only the count, the name should reflect this, eg: `getNodeLinksCount()`

dee077 replied:
changed as requested

nemesifier commented on test/browser.test.utils.js:
```diff
@@ -0,0 +1,96 @@
+  }
+};
+
+export const getPresentNodesAndLinks = async (example) => {
```
same here

dee077 replied:
fixed

nemesifier reviewed (changes requested):
Looks good, but the build is failing. I tried running the tests locally and I get the same error.

Can you add a test for the basic example please? With that we can be relatively safe when upgrading echarts.

We have selenium tests in nodeJS in https://github.com/openwisp/openwisp-wifi-login-pages, you can take a look there for reference.

dee077 commented:
## Changes
The build failed because the Chrome driver couldn't locate `localhost:8080` in the test. This happened because the server was not running in GitHub Actions. To fix this, I updated `ci.yml` to start the server before running tests and added explicit support for the Chrome driver in the GitHub Actions environment.

### Fixes
- Renamed some functions as requested.
- Added a similar test for the Basic Usage Example.
- Running multiple tests was causing some issues, so instead of clicking a link and switching tabs, I used a direct URL to navigate to the page.

### Doubts
- The` openwisp wifi login pages` tests ran on Firefox, as it comes pre-installed in GitHub Actions. However, Firefox is significantly slower and lacks support for retrieving console logs, so I have gone for Chrome.
- When the build checks failed, the error message did not indicate that the `Chrome WebDriver was missing`. Instead, it showed a `connection failure`. This suggests that Chrome and the Chrome WebDriver might have already been installed by default inside the github action server not sure but I have still added it `ci.yml`.

nemesifier reviewed (commented):
Nice progress @dee077! I did a quick review of the code and approved the workflow (waiting for it), see my comments below.

dee077 commented on test/netjsongraph.browser.test.js:
```diff
@@ -20,21 +22,35 @@ describe("Chart Rendering Test", () => {
       driver,
       ".ec-extension-leaflet",
-      2000,
+      5000,
```
This is not waiting time it is timeout, while running the test on github server it may take more time. But if it run fast it will not wait.

nemesifier commented:
The --remote-debugging-pipe flag is an alternative to --remote-debugging-port that allows Chromium-based browsers to communicate with debugging clients (like Selenium, Puppeteer, or Chrome DevTools Protocol) over an IPC (Inter-Process Communication) pipe instead of a TCP port.

I couldn't make it work without this flag!

nemesifier reviewed (approved):
Great work @dee077! :clap:
